lib/puppet/util/windows/api_types.rb

```diff
@@ -22,9 +22,11 @@ class ::FFI::Pointer
 
     def self.from_string_to_wide_string(str, &block)
       str = Puppet::Util::Windows::String.wide_string(str)
-      FFI::MemoryPointer.new(:byte, str.bytesize) do |ptr|
+      FFI::MemoryPointer.new(:byte, str.bytesize + 2) do |ptr|
         # uchar here is synonymous with byte
         ptr.put_array_of_uchar(0, str.bytes.to_a)
+        ptr.put(:uchar, str.bytesize, 0)
+        ptr.put(:uchar, str.bytesize + 1, 0)
```

Comment:
should create a version of `FFI::MemoryPointer.from_wide_string` that behaves like `from_string` and works if a block is given or not.

Reply:
Or perhaps subclass MemoryPointer?

```
class WideStringPointer < MemoryPointer
  def initialize(blah, size)
    super(blah, size + 2)
    put(:uchar, size, 0)
    put(:uchar, size + 1, 0)
  end
end
```